### ISSUE TITLE
Best practice on parallel + incremental / generational ?

### ISSUE THREAD

**Issue by plajjan** [ivmai/bdwgc#545]
I'm looking into using the combination of parallel marking and the incremental GC. I'm primarily looking to reduce the impact of a large heap. I get good performance for small programs where the heap is in MB range. If my heap is multiple GB, every GC pause is many seconds. By detecting only changed pages, I hope to reduce GC pauses considerably by avoiding to scan the entire heap. Instead, the pause time should be more a function of how much temporary memory (i.e. garbage) has been allocated.

I'm using the following options
```
-DLARGE_CONFIG -DGC_BUILTIN_ATOMIC -DGC_THREADS -DNO_PROC_FOR_LIBRARIES -DREDIRECT_MALLOC=GC_malloc -DIGNORE_FREE -DNO_VDB_FOR_STATIC_ROOTS -DPARALLEL_MARK -DGC_TIME_LIMIT=999999
```

I was under the impression that parallel marking would be used when GC_TIME_LIMIT=GC_TIME_UNLIMITED (999999) but my collection runs tell me the parallel marker is disabled:

```
--> Marking for collection #503 after 76986864 allocated bytes
Parallel marking disabled due to stop_func != GC_never_stop_func
Marked from 35791 dirty pages
Stopped marking done after 52883 iterations with disabled parallel marker
World-stopped marking took 59 ms 827197 ns (24 ms in average)
GC #503 freed 140464 bytes, heap 384404 KiB (+ 11032 KiB unmapped + 89550 KiB internal)
In-use heap: 85% (113878 KiB pointers + 214440 KiB other)
0 finalization entries; 0/0 short/long disappearing links alive
0 finalization-ready objects; 0/0 short/long links cleared
Finalize and initiate sweep took 0 ms 1503 ns + 1 ms 437205 ns

--> Marking for collection #504 after 75951824 allocated bytes
Parallel marking disabled due to stop_func != GC_never_stop_func
Marked from 35134 dirty pages
Stopped marking done after 52202 iterations with disabled parallel marker
World-stopped marking took 63 ms 898779 ns (24 ms in average)
GC #504 freed 44400 bytes, heap 370244 KiB (+ 25192 KiB unmapped + 89550 KiB internal)
In-use heap: 88% (114866 KiB pointers + 213568 KiB other)
0 finalization entries; 0/0 short/long disappearing links alive
0 finalization-ready objects; 0/0 short/long links cleared
Finalize and initiate sweep took 0 ms 2073 ns + 2 ms 924925 ns
```

I added extra debug output to determine why it is disabled, which is what we see here:
```
Parallel marking disabled due to stop_func != GC_never_stop_func
```

I don't understand the logic around stop_func. What happens that disables the parallel marker?

Improving program responsiveness is a plus too but picking between 1000 GC pauses each 1ms (better responsiveness) and one GC pause of 500ms (i.e. lower total pause time), I prefer to have fewer and somewhat longer pauses as to reduce total pause time. However, there is an upper limit to what is acceptable here as well.

**Comment by plajjan** [comment]
I suppose one thing I'm interested in understanding is what sort of overhead one can expect from the incremental collector and its various VDB options?

Using parallel mark I get a significant performance improvement, like a test program goes from close to 13 minutes to running in less than 4. However, if I instead enable incremental collection I end up at around 10 minutes, i.e. ~3 minutes better. Using incremental + parallel (with GC_TIME_LIMIT=GC_TIME_UNLIMITED) I end up around 11 minutes, so worse than just the incremental and a lot worse than just parallel. I was really hoping that detecting dirty pages and only scanning them would be a lot more efficient than scanning whole heap. Am I doing something wrong? What are the general expectations when it comes to parallel + incremental? Is it supposed to be fast? :) Am I tuning it wrong?

**Comment by ivmai** [comment]
I haven't seen any benchmarking of parallel mark + incremental/generational yet.

Probably Unity folks did benchmarking (https://blog.unity.com/technology/feature-preview-incremental-garbage-collection, e.g. @jechter).

/cc @hboehm

**Comment by ivmai** [comment]
> I haven't seen any benchmarking of parallel mark + incremental/generational yet.

And nobody tried to optimize it I suppose.

**Comment by plajjan** [comment]
@ivmai let's say I don't care about interrupting a collection after some delay. I can run a "complete collection" but I want it to run through only the dirty pages instead of the whole heap. And I want to run through with parallel markers. Shouldn't this be possible today? That's what I understood -DPARALLEL_MARK + incremental mode with GC_TIME_LIMIT=GC_TIME_UNLIMTED should do? Maybe it does, but it is much slower than plain PARALLEL_MARK. Is this expected?

**Comment by ivmai** [comment]
> I can run a "complete collection" but I want it to run through only the dirty pages instead of the whole heap.

No, this works differently (as explained in [gcdescr.md](https://github.com/ivmai/bdwgc/blob/master/docs/gcdescr.md#generational-collection-and-dirty-bits): the collector may run for a small amount of time not stopping other threads, periodically, and later performs stop-the-world collection of dirty pages only (thus each collection is shorter but totally the collection work takes longer time because the collector needs to rescan dirty pages). Thus, according to your description what you need, the incremental/generation mode is not beneficial for you.

**Comment by plajjan** [comment]
For myself and possibly others. I noticed that disabling huge pages improves performance of the generational GC a lot! Naturally with larger (huge!) pages, a dirty mark means we need to scan more - it seems about 50-60x more for me, so disabling huge pages is a big performance boost. In my test measurements (I'm not really sure how representative of real workloads they are) I have seen overall total runtime increase like 3x by enabling the generational collector, so even though it can achieve somewhat shorter pauses, I still get so much overall time spent in collection. Disabling huge pages though, it is, at least for some test workloads a net gain, simply because we avoid scanning so much of the heap. The larger the live heap is, and the more expensive it is to scan, the better. For example, how things are interconnected, like if we have lots of pointers that references different memory so we end up with a "dense spaghetti ball" of memory, it is seemingly more expensive to scan and thus the generational collector has a greater benefit.

Also beware that the collector grows the heap in increments of 16MB, so for these large tests where I run with heaps 5GB - 20GB, a normal collector would spend a very large amount of time growing the heap in increments of 16MB and forcing a full collection every time. I explicitly set the initial heap size in my testing to get past this. I think the GC would benefit tremendously from being able to grow the heap in much larger increments.
